### PR TITLE
Plugin events

### LINKED ISSUE #2186: Plugins: event handling
## Is your feature request related to a problem? Please describe 👀

The plugins need to be able to interact with the rest of the app. I'd prefer this to be indirect, such as an event system

## Describe the solution you'd like 🎁
Plugins to be able to subscribe to a set of defined events. The event would need to have
- a type
- record type
- record id
- potentially some data as well

Plugins should also be able to raise events - an example of which is the editing of data which then informs the app that the data state is dirty, or that validation has failed.

The event shape would be the same as the above.

We'd need either a context or zustand state in order to share events between the app and plugin contexts.
I think a one-way event system is sufficient - no need for queuing or acknowledgement at this stage.

### Describe alternatives you've considered 💭
plugins receive a data object, a function could be passed along with that - I don't like how concrete that is, the approach feels brittle. and how to interact in the other direction?

### Moneyworks Jobcode 🧰

OMS:PLUG

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2186 

# 👩🏻‍💻 What does this PR do? 
 
Adds some methods to the `PluginProvider`

- addEventListener
- dispatchEvent
- removeEventListener

The methods require the `EventType` union type to be specified and the type of plugin, which together determine where in the app an event should be raised and handled.

# 🧪 How has/should this change been tested? 

There's an example in the plugin example branch of events - where the events are used in the StockDonor plugin to trigger validation on the stock line edit form, and to save data when the line is saved.

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

**Comment by fergie-nz** [comment]
I'm curious - I can't see these files in the omSupply repo. 

Is this a separate semi-detached code base?

File: client/packages/common/src/plugins/hooks/usePluginColumns.ts
Comment (mark-prins) [inline client/packages/common/src/plugins/hooks/usePluginColumns.ts:13]: using `usePluginProvider()` causes the provider and all children to re-render, so a specific fetch is required.

File: client/packages/common/src/plugins/hooks/usePluginColumns.ts
Comment (mark-prins) [inline client/packages/common/src/plugins/hooks/usePluginColumns.ts:21]: if the same column is added again, we only want one of them! I've seen this with HMR, but it's possible it will happen under normal conditions

File: client/packages/common/src/plugins/hooks/usePluginEvents.ts
Comment (mark-prins) [inline client/packages/common/src/plugins/hooks/usePluginEvents.ts:4]: added this as I was having a lot of re-render issues; and found the need to use a selector function to get values out of zustand state clumsy. Am worried that people won't use it correctly and will have render issues

File: client/packages/common/src/ui/layout/tables/DataTable.tsx
Comment (mark-prins) [inline client/packages/common/src/ui/layout/tables/DataTable.tsx:99]: the column wasn't showing - likely after I'd fixed a needless re-render

File: client/packages/common/src/ui/layout/tables/components/index.tsx
Comment (mark-prins) [inline client/packages/common/src/ui/layout/tables/components/index.tsx:35]: added this, so that I could make use of the basic cell layout without recreating it

File: client/packages/system/src/Stock/Components/StockLineEditModal.tsx
Comment (mark-prins) [inline client/packages/system/src/Stock/Components/StockLineEditModal.tsx:68]: pulled the plugins up here, because the form is mutating the `draft` object on every keypress. the plugin, I decided, doesn't need to have the current values, and is ok with the original value of the stockline object

File: client/packages/system/src/Stock/Components/StockLineEditModal.tsx
Comment (mark-prins) [inline client/packages/system/src/Stock/Components/StockLineEditModal.tsx:89]: adding this so that the validation is updated when the plugin changes it's own data

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (mark-prins) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:78]: added some debounce, as the form is jumpy for me - the mutating of `draft` is causing a render which is noticeable when a plugin is rendered too

**Comment by mark-prins** [comment]
> Is this a separate semi-detached code base?

It's in a feature branch for now - waiting until this one is merged before it's ready to bring into develop

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
There is still some re-render bug. Pressing "d" quickly sends you to the Dashboard:

https://github.com/openmsupply/open-msupply/assets/88299195/ac406765-aca9-4437-8d79-7809bf29a3ac

I actually see some similar bugs for the patient search on the contact-tracing page so it might be some underlying issue.
I set to "Request changes" but it might be another issue. Otherwise it all looks good!

One thing I was wondering about is if we should make it more explicit which events are valid for which `ComponentPluginType`. For example, we could do some TS magic to restrict the event type based on the `ComponentPluginType` or we could define events for every `ComponentPluginType` like "StockEditFormSaved", "StockEditFormChanged". 

Another thing I was thinking about is a seperation between events that are coming from plugins and for events that are comming from the host. We could then only make events that are applicable to plugins visible to the plugin, i.e. which events they can listen to and which event they can subscribe to. This might help when implementing plugins, e.g. don't accidentally send the wrong event which then affects other plugins... One way it could be done is like https://developer.chrome.com/docs/extensions/reference/events/.

Regarding security, we more or less fully trust plugins, don't we? so we don't have to restrict what events a plugin can send, right? For example, a plugin could send an event for a different plugin type.

File: client/packages/common/src/plugins/types.ts
Comment (mark-prins) [reply client/packages/common/src/plugins/types.ts:40]: I thought of events as more flexible than that - either the host or the plugin is able to either subscribe to and raise events.

File: client/packages/system/src/Stock/Components/StockLineEditModal.tsx
Comment (mark-prins) [reply client/packages/system/src/Stock/Components/StockLineEditModal.tsx:89]: the `onCancel` is vaguely what I had in mind for that, it's slightly different than unchanging tho. I couldn't see a need to 'unChange' yet, and I've only implemented the minimum of what I think is useful

File: client/packages/common/src/plugins/hooks/usePluginEvents.ts
Comment (mark-prins) [reply client/packages/common/src/plugins/hooks/usePluginEvents.ts:4]: The advantage is usually talked about in terms of ease of use. personally, I've found the need to selectively extract rather annoying to use.

File: client/packages/common/src/plugins/types.ts
Comment (clemens-msupply) [reply client/packages/common/src/plugins/types.ts:40]: Ok might be able to use TypeScript to restrict the EventType base on the plugin area...

Some docs comment about the event types would be useful. For example, `onChange` is confusing to me this is like a command isn't it, like `markAsDirty` isn't it?

**Review by clemens-msupply — APPROVED** [review]
Realized, my re-render complain is actually about the example plugin and not really part of this PR is it? so will approve.

File: client/packages/common/src/ui/layout/tables/components/index.tsx
Comment (mark-prins) [reply client/packages/common/src/ui/layout/tables/components/index.tsx:35]: In the `StockDonorColumn` I'm using the same cell layout and rather than just duplicating that, I've used the `BasicCellLayout` which is exported here, so:

```
  return <BasicCellLayout>{data?.data ?? ''}</BasicCellLayout>;
```

I wanted the same border radius and padding and text overflow that the basic cell has, so that the new cell looks the same. I also want the plugin cell to be updated if the base properties ever get changed

File: client/packages/common/src/plugins/components/PluginProvider.tsx
Comment (mark-prins) [reply client/packages/common/src/plugins/components/PluginProvider.tsx:98]: I've followed the same pattern that javascript uses for events, which is that the listener is only removed if it is an exact match for the object which was added. I thought that this was less limiting than only allowing a single listener for a given combination of eventType and pluginType.. though in practice I think that a single listener is a valid approach.

I'll change to use PluginEventListener and use the types as the match when adding and removing. I think that'll be clearer, thanks

**Comment by mark-prins** [comment]
I fixed the re-render issue; it is a problem with the modal not the plugin. The issue is with the `onUpdate` which was causing a lot of renders - when the draft object isn't changed, but the method is called, then a new object is created and the setState causes a render. It's a little clumsy, but to resolve I've added an `isEqual` check.

The problem surfaces because the cost price input has `autoFocus` set, when you click into another input, this fires the `onChange` event for the cost price, calling `onUpdate` and making the form render about 20 times. This is resolved now.

**Comment by mark-prins** [comment]
thanks for the review @clemens-msupply!

I've changed the event types to be very specific types (onSaveStockEditForm and onChangeStockEditForm) even though this is not my preference. I would have the more standard `onChange` and `onSave` and have the potential for dispatching events for which a handler has not been implemented. 

The change, I think, addresses the concern about limiting a type of event being dispatched for an incorrect type of plugin - though it does still allow an event to be dispatched for a different plugin (the StockDonorEditInput component could fire an event for 'onChangeOutboundShipmentEditForm' if such a thing existed)

**Comment by clemens-msupply** [comment]
thanks for the changes!

> I've changed the event types to be very specific types (onSaveStockEditForm and onChangeStockEditForm) even though this is not my preference. I would have the more standard `onChange` and `onSave` and have the potential for dispatching events for which a handler has not been implemented.

I am open for it. We just need to document which events can happen for which plugins, etc... Think we can even enforce this using TS, e.g. `addListener<'Dashboard'>(...)`

**Pull request merged**
